### ISSUE TITLE
[4.1][Bug] field and filter names being changed by Backpack

### ISSUE THREAD

**Issue by trip-somers** [Laravel-Backpack/CRUD#3065]
# Bug report

### What I did

Upgraded from v3.6 to v4.1 by following instructions in documentation.

### What I expected to happen

Expected behavior similar to v3.6 where field and filter names matched the 'name' attribute of the field/filter definition.

### What happened

(1) Field names are being array-ified based on the 'entity' attribute (I think?). The following example creates a field named `series[line][manufacturer_id]` instead of a field named `manufacturer_id`. Because the field is not named `manufacturer_id`, the dependent ajax filter returns 0 results.

```php
            [
                'tab' => 'Basics',
                'name' => 'manufacturer_id',
                'label' => 'Manufacturer',
                'type' => 'select2',
                'entity' => 'series.line.manufacturer',
                'attribute' => 'name',
                'options' => (function ($query) {
                    return $query->orderBy('name')->get();
                }),
            ],
```

(2) Filter names are being camelCased based on the 'name' attribute. The following example creates a filter named `manufacturerId` instead of a filter named `manufacturer_id`. The downstream effect is similar to the renamed field example when dependent filters are used.

```php
        $this->crud->addFilter(
            [
                'name' => 'manufacturer_id',
                'type' => 'select2',
                'label'=> 'Manufacturer',
            ],
            function() {
                return \App\Catalog\Models\Manufacturer::orderBy('name')->get()
                    ->pluck('name', 'id')->toArray();
            },
            function ($value) { // if the filter is active
                if (! is_null($value)) {
                    $this->crud->addClause('whereHas', 'series', function ($series) use ($value) {
                        return $series->whereHas('line', function ($line) use ($value) {
                            return $line->where('manufacturer_id', $value);
                        });
                    });
                }
            }
        );
```

### What I've already tried to fix it

I've read all of the issues and pull/merge requests I can find about these two similar changes. They appear to have been done on purpose, but I can not find an explanation.

### Backpack, Laravel, PHP, DB version

When I run ```php artisan backpack:version``` the output is:

```sh
### PHP VERSION:
PHP 7.3.5-1+ubuntu18.04.1+deb.sury.org+1 (cli) (built: May  3 2019 10:00:24) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.5, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.5-1+ubuntu18.04.1+deb.sury.org+1, Copyright (c) 1999-2018, by Zend Technologies

### LARAVEL VERSION:
v7.21.0@3ccdb116524de408fdc00715b6f06a1031ddace9

### BACKPACK VERSION:
4.1.15@6c751de946a9c8511dd32eb7bfa3ca6a568849f5
```

**Comment by tabacitu** [comment]
@pxpm will know more about this, for sure. But until he gets to this, I'll try to help:

- (1) I believe in FIELDS it's been done to allow for dot notation names, not just for 1-1 relationships but any type of relationship. In Backpack 4.1, upon `addField()` backpack goes a but further and tries to guess things for you, so that you don't have to specify _everything_ for a field. For relationships especially, this is useful because most of the times you can just specify the `name` and it'll figure out the rest. In this particular use case, I think what's happening is that your `manufacturer_id` gets overwritten by the `addField()` statement with `series[line][manufacturer_id]` because it sees a chained relationship inside `entity`. And I see we might not have a way to tell Backpack "_please don't overwrite my name, I know what I'm doing here_". @pxpm correct me if I'm wrong and we DO have a way to avoid this - I remember we talked about this at some point but can't remember the resolution. Anyway, to fix this issue for you and others that have the same problem, if we don't already have a solution, I propose we add another optional attribute, something like `overwrite_name` that you could set to `false`, and the Backpack would know NOT to touch your name; so something like this inside `FieldsProtectedMethods.php` - what do you think @trip-somers  & @pxpm ?
![Screenshot 2020-07-24 at 08 15 58](https://user-images.githubusercontent.com/1032474/88362637-f56b0f00-cd85-11ea-84e0-312c269153cf.png)




- (2) I believe in FILTERS it's been done to fix a small bug where the "Remove filters" button didn't show every time (#2829 and #2872), and for consistency - because throughout the filters we were using sometimes camelCase, sometimes snake_case versions of the input name, and this created random problems with some names.

Note - also related and relevant, your [comment here](https://github.com/Laravel-Backpack/CRUD/pull/2872#issuecomment-663332716).

This must've ruined your experience when upgrading to 4.1 @trip-somers - sorry for that. Looks like this change was very badly documented. It might be the case that we've over-optimized here - looks like when trying to fix a minor bug we introduced a major breaking change... 👀  And a badly documented one... Let's see how we can best fix this... It shouldn't be that difficult to upgrade dependant selects / dependant filters, or we should at least provide clear docs for it...

**Comment by trip-somers** [comment]
My experience has definitely not been bad at all. I'm not sweating these issues right now either. The helpful and thoughtful responses are a breath of fresh air in the open source community at large that seems to frequently close issues without comment and get defensive or even combative.

For my filters, it was slightly annoying but relatively easy to adjust my filter names to match how they'd be transformed. If you can point me to a relevant thread, I can post the code for my dependency-based filter. Seems small enough and easy to be included in the standard `select2_ajax` filter. I figure it's a common enough extension pattern that I'm not the only one affected, and including it in CRUD would elevate it from an extension to a feature.

The fields issue threw me for a loop because I needed to add a `\Log::debug()` to see what Backpack was even sending to my API (visually parsing the url-encoded query string was a nope for me). I was plenty surprised to see that `manufacturer_id` was somehow `series[line][manufacturer_id]`, and I didn't want to start hacking my API to expect that without knowing how or why it was being changed.

Happy to be a part of any discussions related to fixes or adjustments.

**Comment by trip-somers** [comment]
Quick question about the `overwriteFieldNameFromEntity()` method: how would this actually work in practice? When parsing the `form` request data as suggested in the FAQ, you wouldn't be able to use dot-notation to retrieve the value.

```php
$form = collect($request->input('form'))->pluck('value', 'name');
```

This would result in `$form['series[line][manufacturer_id]']` rather than something like `$form['series.line.manufacturer_id']`, wouldn't it? Does it work different when the field is posted for create or update?

**Comment by trip-somers** [comment]
Encountered another side effect related to the field name change issue. I'll make the example as short as possible, building from what has been mentioned above...

* If I use 'manufacturer_id' as the first of a series of `select2_from_ajax` dependencies in the Item form, the 'manufacturer_id' field name winds up being `series[line][manufacturer_id]`.
* If I use 'manufacturer_id' as the first of a series of `select2_from_ajax` dependencies in the Series form, the 'manufacturer_id' field name winds up being `line[manufacturer_id]`.

This means the Line API Controller has to check for 2 fields other than the obvious `manufacturer_id` when processing a search request. If the field name isn't fixed, it makes API development... messy.

**Comment by pxpm** [comment]
Hello @trip-somers and @tabacitu 

Indeed this is a different scenario from the filters. 

From screenshot above from @tabacitu I already saw a bug, introduced by me ofc. :smiling_imp: Sorry for that. 

It reads: `We should only change the 1-1 relation field names ...` but actually we check `relation_type` equals to `BelongsTo` instead of `HasOne`.

 In `HasOne`, a relation that makes sense you could edit in the same crud, we overwrite the name of the field as an array to allow us to have for example a field with name `type` on the main crud, with an HasOne relation with something that has also a `type` field name. Also for validation you validate as a Laravel array: `relation.type => required` 

Again, sorry for introducing such bad experience. 

Backpack is used in so many different ways that sometimes it gets very difficult to test every possible scenario. 

I am working on a local branch in a way to solve this naming issues. 

In the meanwhile change line 180 of FieldsProtectedMethods.php from `BelongsTo` to `HasOne`. It should solve your select issues.

Let me know,
Pedro

**Comment by trip-somers** [comment]
Thank you for checking this out. I will try it out tonight.

**Comment by trip-somers** [comment]
As you suggested, changing line 180 fixed the field name issue; however, it appears to have revealed that the actual ID is not being sent.

Instead of `series[line][manufacturer_id]`, the field is now called `manufacturer_id` as expected, but instead of `8`, the form value is `Widget Manufacturer`.

**Comment by trip-somers** [comment]
Can I get an update on the status of the select2 field dependencies issues? Has the dependency value issue from my last comment been acknowledged?

Are you aiming for 4.2 or a 4.1.x release?

**Comment by pxpm** [comment]
@trip-somers hello. 

Hope to find you well. 

About the filters I think we already got to a consensus but will be pushed as a BC in 4.2. 

About the fields I think there is something we can do in 4.1.

Correct me here if I am wrong, the only reason why you are using `'entity' => 'series.line.manufacturer',` is to don't write `'model' => '\App\Catalog\Models\Manufacturer'` right ? 

Can you post me the two dependant fields that are not working for you ? Also the relations types. 

I could not reproduce the last error you report about `Widget Manufacturer`.

This are my fieds:

```php

$this->crud->addField([
            'name' => 'number',
            'type' => 'select2',
            'attribute' => 'name',
            'entity' => 'category',
        ]);

$this->crud->addField([
            'name' => 'select2_from_ajax',
            'type' => 'select2_from_ajax',
            'data_source' => backpack_url('features/fetch/article'),
            'dependencies' => ['number'],
            'include_all_form_fields' => true,
            'placeholder' => 'select smth',
            'minimum_input_length' => 0,
            'method' => 'POST',
            'entity' => 'article',
            'tab' => 'Relations',
            'label' => 'select2 from ajax depends on category'
        ]);
```

I am setting up a panel with pretty much all supported relations and some advanced use cases like `dependant fields`.  Apart from the renaming of the `BelongsTo` relations that should not happen I didn't found anything else broken for now.

Also you are using FetchOperation for your ajax routes or you setup your own API routes ? Could you make the change we discussed earlier and it's on PR: #3116 and guide me throw the rest of your error report? 

Best,
Pedro

**Comment by trip-somers** [comment]
Regarding 'entity' vs 'model' -- I'm using 'entity' because the documentation for select2_from_ajax implies that 'entity' is required and 'model' is optional. There is nothing the documentation that I've seen that indicates why someone would choose one or the other, or if using both would be weird.

I am not using the `FetchOperation`. I am using the structure suggested in the documentation here: https://backpackforlaravel.com/docs/4.1/crud-how-to#add-a-select2-field-that-depends-on-another-field

It makes more sense for me to do that since 'manufacturer_id' field is used in 3 places and 'line_id' is used in 2 places. But also, I did not know that `FetchOperation` had been added.

Testing this morning shows the value being the `id` field instead of `name`. I know I was consistently receiving `Widget Manufacturer` instead of `8` when playing with this two weeks ago, but now I am unable to reproduce it. I will dig into that again tonight to see if I can figure out what I did.

**Comment by trip-somers** [comment]
I was intrigued so I looked at this now, and it's happening in one place but not another.

In my ItemCrudController, which works just fine:

```php
            [
                'tab'                   => 'Basics',
                'name'                  => 'manufacturer_id',
                'label'                 => 'Manufacturer',
                'type'                  => 'select2_from_ajax',
                'entity'                => 'series.line.manufacturer',
                'attribute'             => 'name',
                'data_source'           => url('admin/api/manufacturer'),
                'placeholder'           => '- Select Manufacturer -',
                'minimum_input_length'  => 0,
            ],
            [
                'tab'                       => 'Basics',
                'name'                      => 'line_id',
                'label'                     => 'Line',
                'type'                      => 'select2_from_ajax',
                'entity'                    => 'series.line',
                'model'                     => 'App\Catalog\Models\Line',
                'attribute'                 => 'name',
                'data_source'               => url('admin/api/line'),
                'placeholder'               => '- Select Line -',
                'minimum_input_length'      => 0,
                'dependencies'              => ['manufacturer_id'],
                'include_all_form_fields'   => true,
            ],
```

From my SeriesCrudController, which definitely is sending 'Widget Manufactuer' as `manufacturer_id`:

```php
            [
                'name'                  => 'manufacturer_id',
                'label'                 => 'Manufacturer',
                'type'                  => 'select2_from_ajax',
                'entity'                => 'line.manufacturer',
                'attribute'             => 'name',
                'data_source'           => url('admin/api/manufacturer'),
                'placeholder'           => '- Select Manufacturer -',
                'minimum_input_length'  => 0,
            ],
            [
                'name'                      => 'line_id',
                'label'                     => 'Line',
                'type'                      => 'select2_from_ajax',
                'entity'                    => 'line',
                'model'                     => 'App\Catalog\Models\Line',
                'attribute'                 => 'name',
                'data_source'               => url('admin/api/line'),
                'placeholder'               => '- Select Line -',
                'minimum_input_length'      => 0,
                'dependencies'              => ['manufacturer_id'],
                'include_all_form_fields'   => true,
                'wrapperAttributes' => [
                    'class' => 'form-group col-lg-6'
                ],
            ],
```

The only appreciable difference I can see are the 'entity' values, but I'm not sure why or how that would result in `'manufacturer_id' => 'Widget Manufacturer'` instead of the expected `'manufacturer_id' => 8` in the `line_id` ajax request.

**Comment by trip-somers** [comment]
I am in the very weird position of can not fix and can not duplicate. I have moved fields, changed field names, and varied combinations of `entity` and `model` usage. I have tried to fix it on SeriesCrudController without luck. I have tried to _cause_ it on ItemCrudController without luck.

I have confirmed that the two pages make equivalent ajax requests and receive equivalent responses, but somehow ItemCrudController is assigning the ID as the selected value -- `<option value="8" data-select2-id="31">Widget Manufacturer</option>` -- while SeriesCrudController is assigning the name -- `<option value="Widget Manufacturer" data-select2-id="28">Widget Manufacturer</option>`.

It makes no sense. I am at a loss.

**Comment by trip-somers** [comment]
I finally found it. I built a custom field called `select2_from_ajax_plain` that is designed to work with non-relationship. I extended it from `select2_from_ajax` but failed to change the field init function name. The result was that my new `select2_from_ajax_plain` field was breaking the init function for the `select2_from_ajax` fields on the page.

So this is back down to just the BelongsTo/HasOne mistake. My apologies if you spent much time trying to recreate this. I'm a unicorn.

**Comment by pxpm** [comment]
Hey @trip-somers 

Glad you found it.

My notes from friday: 

```
....
- could not reproduce wrong sent of value using api or fetch operation. 
- could be environment specific ? Ask for dev. environment.
- ask if user had overwritten any of the field files.
```

I was also getting into the last possible scenarios.

Don't mind the time I spent debugging this, I found more stuff that we can optimize, and built a complete relation panel to test further issues if needed, it's not a waste of time. 

From this issue I think there is not much more to do here, filter names are going to be addressed in 4.2 as a BC, and I think #3116 has the fix for the issue of BelongsTo beeing renamed. 

Thanks again for submitting the issue, if you found some more let us know so we can properly address it. 

**FYI:** you can use unrelated ajax fields. We should update our docs accordingly, sorry for misdirecting you. 

```php
 $this->crud->addField([
            'name' => 'select2_from_ajax',
            'type' => 'select2_from_ajax',
            'data_source' => backpack_url('features/fetch/article'),
            'dependencies' => ['number'],
            'include_all_form_fields' => true,
            'placeholder' => 'select smth',
            'minimum_input_length' => 0,
            'method' => 'POST',
            //'entity' => 'article',
            'model' => 'Backpack\NewsCRUD\app\Models\Article',
            'attribute' => 'title',
            'tab' => 'Relations',
            'label' => 'select2 from ajax depends on category'
        ]);
```
It works just like removing model and adding `entity => 'article'`, the difference is that if entity is set, we infer the model from the relation and some other attributes.

Best,
Pedro

**Comment by trip-somers** [comment]
Gave it a shot, and it still requires at least a model. My ajax request is a bit more abstract than that -- querying the parent Line  for a list of Groups that a Series can be assigned to. The Groups data exists as a json field on the Line rather than as its own model.

What I've got works, so I'll stick with it for now.

**Comment by tabacitu** [comment]
Hi guys,

Whew. It looks like this thread has advanced quite a bit. So much so that it's a bit difficult to follow. 
So let me sum it up - please let me know if I've got it right/wrong.

# Fields

###  ✅ Problem 1 A - field names were array-ified for `BelongsTo` instead of `HasOne` 
- fixed by @pxpm in #3116
- needs to be tagged as Backpack 4.1.18 for everyone to get the fix;

### 🟨 Problem 1 B - no way to tell Backpack "_don't overwrite my field name_" 
- mentioned by @tabacitu in https://github.com/Laravel-Backpack/CRUD/issues/3065#issuecomment-663351403 
- what do you think @pxpm & @trip-somers - is this still needed? wouldn't hurt though, right? I mean we should probably provide a way for the developer to take complete control over this (aka disable the new behaviour in 4.1 where we overwrite the field name), even if it's not exactly this solution;

### ✅ Problem 1 C - Getting `Widget Manufacturer` instead of `8` as the value of the field 
- turns out it was user error, due to a custom field (https://github.com/Laravel-Backpack/CRUD/issues/3065#issuecomment-674478286)

### ✅ Problem 1 D - Docs are unclear/wrong in saying that `entity` is mandatory
- additionally, they don't make it clear if you should use `model` or `entity`;
- I've worked on the docs a little bit, hopefully this provides all the information one needs to understand how fields containing relationships work; see https://github.com/Laravel-Backpack/docs/pull/171

# Filters

### 🟨 Problem 2 - filter names get camelCased
- @pxpm mentioned in https://github.com/Laravel-Backpack/CRUD/issues/3065#issuecomment-674759976 that we decided to fix this in 4.2 as a breaking change, but actually recently we decided it's best to fix this in 4.1, right way, so that fewer people get run into it; https://github.com/Laravel-Backpack/CRUD/pull/3069#issuecomment-674999998 gives more details; I'd rather we continue the conversation about this there, since it's already work-in-progress;

----

Bottom line, they will both get fixed in 4.1. Probably by the end of the week.

----

Let me know if I've misunderstood / missed something @trip-somers & @pxpm . I'm eager to finally see this closed, after we settle the 2 remaining yellow items above. It's been bugging me for past month, knowing we have these problems in our code.

Cheers!

PS. On the bright side @pxpm the demo you've been working on (that treats ALL relationship types and combinations) will be very useful in the future. I'll help a lot with maintenance and confirming/disproving bug reports - I'm 100% sure of it 😄 🥳

**Comment by trip-somers** [comment]
Everything looks good to me. My comment for 1B is that if the plan is to continue to transform HasOne field names, then there should be some kind of override for the developer or at least some kind of callout/warning in the docs that use of 'entity' will transform the field name of a HasOne relationship.

^ And now that I've written that out and read it back to myself, we're talking about a quirk that only effects 'HasOne' field names defined with an 'entity' value. Is it also limited to 'select2_from_ajax' field types? I don't know why that behavior was added, but maybe it just makes more sense to remove it since it has such a narrow scope. If the developer wants the field name to be array-ified, it seems like that should be the developer's call.

**Comment by tabacitu** [comment]
@trip-somers granted it's a narrow scope. What it does (not only if `entity` is present) is that it allows Backpack developers to use easily use 1-to-1 relationships, by just writing out a the relationship as a `name`, using dot notation.

So for example:
```php
CRUD::field('user.details.emergency_contact');
```

This will add an input with the name `details[emergency_contact]` inside the form. That will automatically get saved thanks to the 1-to-1 `details()` relationship on the User model. 

Without this feature, the only way to do it would have been to do:
```php
CRUD::field('emergency_contact');
```
and either setup an accessor & mutator for it, or overwrite the save process to add this bit. So quite a lot more work.

Personally I never use 1-to-1 relationship, but quite a few people do. And since all other relationships worked fine inside Backpack, we were basically discriminating against 1-to-1 relationships 😅

The choice to transform the name from dot-notation to bracket-notation came because of inconsistencies between browsers - some support dot notation, others don't. So out of need.

I'm not happy with the unintuitive behaviour either, but I think it's best to provide support for 1-to-1 relationships too, and I hope we can fix this problem with better docs 🤞

**Comment by trip-somers** [comment]
That's a good explanation. I wasn't aware that short-hand even existed.

**Comment by tabacitu** [comment]
Separated issue 1B (the only one that's not done at the moment) into https://github.com/Laravel-Backpack/CRUD/issues/3150 so let's close this 🥳 🎉 

Thanks a lot for the time and patience @pxpm and @trip-somers !
